packages/types/src/getters/utils/create-getter.ts

```diff
@@ -0,0 +1,23 @@
+import { Getter } from './getter';
+
+export const createGetter = <SourceType, TargetType, AllowsUndefined extends boolean = true>(
+  getterFunction: (
+    value: SourceType | undefined,
+  ) => AllowsUndefined extends true ? TargetType | undefined : TargetType,
+): Getter<SourceType, TargetType, AllowsUndefined> => {
+  const getter: Getter<SourceType, TargetType, AllowsUndefined> = value => getterFunction(value);
+
+  getter.orThrow = errorMessage => {
+    const getOrThrow = createGetter<SourceType, TargetType, false>(value => {
+      const result = getterFunction(value);
+      if (typeof result === 'undefined') throw new Error(errorMessage);
+      return result;
+    });
```

Comment:
Instead of writing our own errors, think we have it write its own:

```typescript
  getter.orThrow = () => {
    return createGetter<SourceType, TargetType, false>(value => {
      const result = getterFunction(value);
      if (typeof result === 'undefined') {
        const errorMessage = `${getterFunction.name} failed to extract from ${JSON.stringify(value)}`;
        throw new Error(errorMessage);
      }
      return result;
    });
  };
```

Reply:
Good call. I added the second half of that error message [here](https://github.com/penumbra-zone/web/pull/484/commits/16c294a79be22d442a310f4b8962d82ae1431f39#diff-0aafb5759e68a96372f36cbab5c7a0347e7a39227ef6b3f3905796047f42a206R12). Note that the getter function has no name, because `createGetter` takes an anonymous function:
```ts
const getFirstName = createGetter((employee: Employee) => employee.firstName);
                               // ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                               // anonymous function
```

We could require that people define function names:
```ts
const getFirstName = createGetter(function getFirstName(employee: Employee) { return employee.firstName) });
```

But that doesn't feel as ergonomic. Thoughts?